### PR TITLE
Port option parsing logic from `dask.dataframe.read_parquet` to `to_parquet`

### LINKED ISSUE #9980: `dask.dataframe` inconsistent in supporting `filesystem` argument
The parquet handling is inconsistent between `read_parquet` and `to_parquet` in that `read_parquet` supports passing an initialized filesystem object, but `to_parquet` does not. This seems to be due to a recent partial refactor of `read_parquet` which streamlined the options processing. 

When passing an `fsspec.AbstractFileSystem` object to `to_parquet`, paths get interpreted as local paths instead of dispatching to the filesystem. This leads to confusing errors. 

It should be relatively straightforward to port the changes from `read_parquet` to fix this.

**Minimal Complete Verifiable Example**:

```python
import fsspec
import pyarrow as pa
import pyarrow.dataset as pa_da
import dask.dataframe as dd
import pandas as pd
import numpy as np

# generate some sample data
gen = np.random.default_rng(123)
df = pd.DataFrame(data=gen.standard_normal(size=(5, 3)), columns=["a", "b", "c"])

# write some data to fs using pyarrow dataset api
IN_KEY = "/sample"
OUT_KEY = "/sample2"
fs = fsspec.get_filesystem_class("memory")()
pa_da.write_dataset(pa.Table.from_pandas(df), IN_KEY, format='parquet', filesystem=fs, existing_data_behavior="overwrite_or_ignore")
print(fs.ls(IN_KEY, detail=False))

# read with initialized fs
g = dd.read_parquet(IN_KEY, filesystem=fs)

# write to initialized fs
dd.to_parquet(g, OUT_KEY, filesystem=fs)  # interprets path as local fs; fails with OSError.
```

**Anything else we need to know?**:

Sample stacktrace: 
image

**Environment**:

- Dask version: current main, 0eb4bd0b
- Python version: python-3.10
- Operating System: osx
- Install method (conda, pip, source): conda

### PR DESCRIPTION
**Pull request by antonl**
@rjzamora I may need your help using the extract_filesystem function correctly, but this PR passes my tests. More features could probably be added to intelligently pass options to write_dataset.

- [x] Closes #9980 
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

**Review by jrbourbeau — COMMENTED** [review]
Thanks @antonl

cc @rjzamora

**Comment by antonl** [comment]
Failing tests don't have anything to do with this change: 
image

**Review by jrbourbeau — COMMENTED** [review]
@antonl those failures are unrelated to the changes here and should be fixed by merging `main` into this branch

**Comment by antonl** [comment]
@jrbourbeau Ok, another merge seems to get it in.

File: dask/dataframe/io/parquet/core.py [hunk 1]
```diff
@@ -796,6 +797,10 @@ def to_parquet(
         If not specified, files will created using the convention
         ``part.0.parquet``, ``part.1.parquet``, ``part.2.parquet``, ...
         and so on for each partition in the DataFrame.
+    filesystem: "fsspec", "arrow", fsspec.AbstractFileSystem, or pyarrow.fs.FileSystem
+        Filesystem backend to use. Note that the "fastparquet" engine only
+        supports "fsspec" or an explicit ``pyarrow.fs.FileSystem`` object.
```
Comment (rjzamora) [inline dask/dataframe/io/parquet/core.py:802]: Ah - Looks like there is an error in the `read_parquet` doc-string that you are copying here. Pretty sure this should be:

```
supports "fsspec" or an explicit ``fsspec.AbstractFileSystem`` object.
```

**Review by rjzamora — COMMENTED** [review]
Thanks for the contribution @antonl !  I'd like to expand test coverage a bit, but the general change seems good to me.

File: dask/dataframe/io/parquet/core.py (hunk 1 above)
Comment (antonl) [reply dask/dataframe/io/parquet/core.py]: Fixed!

File: dask/dataframe/io/tests/test_parquet.py [hunk 2]
```diff
@@ -4557,6 +4557,37 @@ def test_pyarrow_filesystem_option(tmp_path, fs):
...
+    key1 = "/test_fsspec_to_parquet_filesystem_option/pa_write"
+    key2 = str(tmp_path / "da_write")
+
+    df = pd.DataFrame({"a": range(10)})
+    fs = get_filesystem_class("memory")()
+    pa_da.write_dataset(
+        pa.Table.from_pandas(df),
+        key1,
+        format="parquet",
+        filesystem=fs,
```
[the remark this answers is not in the document]
Comment (antonl) [reply dask/dataframe/io/tests/test_parquet.py:4574]: I've extended the `test_pyarrow_filesystem_option` to also use filesystem flag to write. Does that address your first concern?

I've extended the test case to also exercise the append=True option.

File: dask/dataframe/io/tests/test_parquet.py [hunk 3]
```diff
@@ -4557,6 +4557,47 @@ def test_pyarrow_filesystem_option(tmp_path, fs):
...
+    assert len(list(tmp_path.iterdir())) == 0, "wrote to local fs"
+
+    # make sure we wrote a key to memory fs
+    assert len(fs.ls(key2, detail=False)) == 1
+
+    # ensure append functionality works
+    ddf.to_parquet(key2, engine="pyarrow", append=True, filesystem=fs)
+    assert len(fs.ls(key2, detail=False)) == 2, "should have two parts"
+
+    rddf = dd.read_parquet(key2, engine="fastparquet", filesystem=fs)
```
Comment (rjzamora) [inline dask/dataframe/io/tests/test_parquet.py:4595]:
```suggestion
    rddf = dd.read_parquet(key2, engine="pyarrow", filesystem=fs)
```
There is no guarantee that `fastparquet` is installed when this test runs.

File: dask/dataframe/io/tests/test_parquet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rjzamora) [reply dask/dataframe/io/tests/test_parquet.py:4574]: >Does that address your first concern?

Yes - thanks!

File: dask/dataframe/io/tests/test_parquet.py [hunk 4]
```diff
@@ -4557,6 +4557,47 @@ def test_pyarrow_filesystem_option(tmp_path, fs):
...
+    key2 = str(tmp_path / "write1")
+
+    df = pd.DataFrame({"a": range(10)})
+    fs = get_filesystem_class("memory")(use_instance_cache=False)
+    pa_da.write_dataset(
+        pa.Table.from_pandas(df),
+        key1,
+        format="parquet",
+        filesystem=fs,
+    )
```
Comment (rjzamora) [inline dask/dataframe/io/tests/test_parquet.py:4575]:
```suggestion

    key1 = "/read1"
    key2 = str(tmp_path / "write1")

    df = pd.DataFrame({"a": range(10)})
    fs = get_filesystem_class("memory")(use_instance_cache=False)
    df.to_parquet(key1, engine="pyarrow", filesystem=fs)
```

Comment (rjzamora) [reply dask/dataframe/io/tests/test_parquet.py:4575]: Pretty sure pandas will pass through the `filesystem` option for you.

File: dask/dataframe/io/tests/test_parquet.py [hunk 5]
```diff
@@ -4557,6 +4557,47 @@ def test_pyarrow_filesystem_option(tmp_path, fs):
...
+    assert len(fs.ls(key2, detail=False)) == 1
+
+    # ensure append functionality works
+    ddf.to_parquet(key2, engine="pyarrow", append=True, filesystem=fs)
+    assert len(fs.ls(key2, detail=False)) == 2, "should have two parts"
+
+    rddf = dd.read_parquet(key2, engine="fastparquet", filesystem=fs)
+    assert_eq(rddf.partitions[0], ddf)
+    assert_eq(rddf.partitions[1], ddf)
+    pd.testing.assert_frame_equal(rddf.partitions[0].compute(), df)
```
Comment (rjzamora) [inline dask/dataframe/io/tests/test_parquet.py:4598]:
```suggestion
    assert_eq(ddf, df)
```
If this check passes (along with the two assertions above), then we know both partitions are the same as `df` after `compute`. Note that this assertion can also be moved to the like just before `ddf.to_parquet` for clarity.

**Review by rjzamora — COMMENTED** [review]
This is looking good @antonl - Only have a few small suggestions.

File: dask/dataframe/io/tests/test_parquet.py (hunk 5 above)
Comment (rjzamora) [reply dask/dataframe/io/tests/test_parquet.py:4598]: Note that you can also replace

```python
    assert_eq(rddf.partitions[0], ddf)
    assert_eq(rddf.partitions[1], ddf)
```
with 

```python
    assert_eq(rddf, dd.concat([ddf, ddf]))
```

File: dask/dataframe/io/tests/test_parquet.py (hunk 4 above)
Comment (antonl) [reply dask/dataframe/io/tests/test_parquet.py:4575]: I like the ability to skip instance cache, so I chose it. I think I could have used `memory:///{key1}`.

File: dask/dataframe/io/tests/test_parquet.py (hunk 5 above)
Comment (antonl) [reply dask/dataframe/io/tests/test_parquet.py:4598]: That makes sense. I wasn't sure whether assert_eq only compares the dask graphs, or actually evaluates the contents.

File: dask/dataframe/io/tests/test_parquet.py (hunk 3 above)
Comment (antonl) [reply dask/dataframe/io/tests/test_parquet.py:4595]: Ah makes sense.

File: dask/dataframe/io/tests/test_parquet.py (hunk 4 above)
Comment (antonl) [reply dask/dataframe/io/tests/test_parquet.py:4575]: Didn't know about this, cool!

**Comment by antonl** [comment]
I keep getting unrelated test failures. 
image

**Comment by j-bennet** [comment]
@antonl If you're getting errors in pyarrow build ("Tests / test (ubuntu-latest, 3.10, pyarrow)"), don't worry about those. They are pointing to nightly builds of `pandas`, `numpy` and `pyarrow`, and if things break there, it's likely not your problem.

**Review by rjzamora — APPROVED** [review]
LGTM - Thanks for the nice contribution @antonl !

**Pull request merged**